Sources/Authentication/Credentials/UsernamePassword.swift

```diff
@@ -0,0 +1,83 @@
+// MARK: Data structure
+
+public struct Password: Crendentials {
+    public let username: String
+    public let password: String
+    public let verifier: PasswordVerifier?
+
+    public init(username: String, password: String, verifier: PasswordVerifier? = nil) {
+        self.username = username
+        self.password = password
+        self.verifier = verifier
+    }
+}
+
+// MARK: Authenticatable
+
+public protocol PasswordAuthenticatable: Authenticatable {
+    // MARK:  Username / Password
+    /// Return the user matching the supplied
+    /// username and password
+    static func authenticate(_: Password) throws -> Self
+
+    /// The entity's raw or hashed password
+    var password: String? { get }
+
+    /// The key under which the user's username,
+    /// email, or other identifing value is stored.
+    static var usernameKey: String { get }
+
+    /// The key under which the user's password
+    /// is stored.
+    static var passwordKey: String { get }
+}
+
+public protocol PasswordVerifier {
+    func verify(password: String, matchesPassword: String) throws -> Bool
```

Comment:
we should never be able to match passwords this way unless our passwords are stored in plaintext, I think it should be `matchesHash: String`